### PR TITLE
fix settings proxying logic and add e2e tests for image element finding

### PR DESCRIPTION
**Pull request by jlipps**
there was a bug where we would proxy the settings update, then do so again during the update callback.

this depends on https://github.com/appium/appium-base-driver/pull/235 so tests will fail and this PR should not be merged until that PR is merged and published.

### CODE REVIEW

File: lib/commands/general.js [hunk 1]
```diff
@@ -3,6 +3,15 @@ import log from '../logger';
 import { androidHelpers } from 'appium-android-driver';
 import { errors } from 'appium-base-driver';
 
+const DRIVER_ONLY_SETTINGS = [
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: I assume this list should not be duplicated here, but rather extended from base driver

File: lib/commands/general.js [hunk 2]
```diff
@@ -109,11 +118,31 @@ commands.openNotifications = async function () {
...
 commands.updateSettings = async function (settings) {
-  return await this.uiautomator2.jwproxy.command('/appium/settings', 'POST', {settings});
+  // we have some settings that are set on the settings object in the driver
+  // only, for example image finding settings. The uiauto2 server does not know
+  // what to do with them, so just set them on this driver's settings instance,
+  // and don't forward them to the server
+  let driverOnlySettings = {};
+  let serverSettings = {};
+  for (let setting of Object.keys(settings)) {
+    if (DRIVER_ONLY_SETTINGS.indexOf(setting) !== -1) {
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: !includes

File: lib/commands/general.js [hunk 3]
```diff
@@ -109,11 +118,31 @@ commands.openNotifications = async function () {
...
 
 commands.updateSettings = async function (settings) {
-  return await this.uiautomator2.jwproxy.command('/appium/settings', 'POST', {settings});
+  // we have some settings that are set on the settings object in the driver
+  // only, for example image finding settings. The uiauto2 server does not know
+  // what to do with them, so just set them on this driver's settings instance,
+  // and don't forward them to the server
+  let driverOnlySettings = {};
+  let serverSettings = {};
+  for (let setting of Object.keys(settings)) {
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: _.keys

File: lib/commands/general.js [hunk 4]
```diff
@@ -109,11 +118,31 @@ commands.openNotifications = async function () {
...
+    if (DRIVER_ONLY_SETTINGS.indexOf(setting) !== -1) {
+      driverOnlySettings[setting] = settings[setting];
+    } else {
+      serverSettings[setting] = settings[setting];
+    }
+  }
+  await this.settings.update(driverOnlySettings);
+  if (Object.keys(serverSettings).length > 0) {
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: !_.isEmpty(serverSettings)

File: lib/commands/general.js [hunk 5]
```diff
@@ -109,11 +118,31 @@ commands.openNotifications = async function () {
...
+  let driverOnlySettings = {};
+  let serverSettings = {};
+  for (let setting of Object.keys(settings)) {
+    if (DRIVER_ONLY_SETTINGS.indexOf(setting) !== -1) {
+      driverOnlySettings[setting] = settings[setting];
+    } else {
+      serverSettings[setting] = settings[setting];
+    }
+  }
+  await this.settings.update(driverOnlySettings);
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: some logging might be useful here

File: lib/commands/general.js (hunk 1 above)
Comment (jlipps) [reply lib/commands/general.js]: yeah good idea

File: lib/commands/general.js (hunk 3 above)
Comment (jlipps) [reply lib/commands/general.js]: why would we prefer lodash over the built-in language feature?

Comment (mykola-mokhnach) [reply lib/commands/general.js]: As for my taste lodash wrappers are safer. They perform safe verification even if passed values are kinda broken

Comment (jlipps) [reply lib/commands/general.js]: fair enough. i also did a little research and we use `_.keys` about twice as much as `Object.keys` in our codebases, so might as well try and keep things consistent

Comment (mykola-mokhnach) [reply lib/commands/general.js]: https://github.com/jashkenas/underscore/issues/149

File: test/functional/commands/find/by-image-e2e-specs.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    await driver.updateSettings({imageMatchThreshold});
+  });
+  it('should be able to get basic element properties', async function () {
+    let el = await driver.elementByImageFile(START_IMG);
+    await el.isDisplayed().should.eventually.be.true;
+    let size = await el.getSize();
+    size.width.should.be.above(0);
+    size.height.should.be.above(0);
+    let loc = await el.getLocation();
+    (loc.x >= 0).should.be.true;
```
Comment (mykola-mokhnach) [inline test/functional/commands/find/by-image-e2e-specs.js]: should be also use `above` here ?

File: lib/commands/general.js [hunk 7]
```diff
@@ -109,11 +109,39 @@ commands.openNotifications = async function () {
...
 
 commands.updateSettings = async function (settings) {
-  return await this.uiautomator2.jwproxy.command('/appium/settings', 'POST', {settings});
+  // we have some settings that are set on the settings object in the driver
+  // only, for example image finding settings. The uiauto2 server does not know
+  // what to do with them, so just set them on this driver's settings instance,
+  // and don't forward them to the server
+  let driverOnlySettings = {};
+  let serverSettings = {};
+  for (let setting of _.keys(settings)) {
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: ID rather perform loop by pairs, since we need both key and value inside the it

**Comment by imurchie** [comment]
See https://github.com/appium/appium-support/blob/master/.travis.yml for configuring install of`opencv4nodejs`

**Comment by jlipps** [comment]
hmm, realizing we really need to merge this for image finding to work on android, despite my inability to get opencv working on travis. @imurchie do you think we should just skip the test for now and merge, and try to get opencv working on linux later?

**Comment by imurchie** [comment]
Sure. I'll give it a shot asynchronously, too.

**Comment by jlipps** [comment]
ok, trying a new travis config

**Comment by jlipps** [comment]
if this doesn't work i'll skip so we can merge

**Comment by jlipps** [comment]
@imurchie @mykola-mokhnach any objections to merging this now?

**Comment by jlipps** [comment]
published in 1.17.1. and my github merge button was not in squash mode on accident. sorry!

**Pull request merged**